Automated holiday generator

tswistak opened the pull request:
<!-- Hey there. Thank you so much for improving Fossify. Please consider filling out the details :)-->

#### What is it?
- [ ] Bugfix
- [ ] Feature
- [X] Codebase improvement

#### Description of the changes in your PR
- What has been done:
  - Node script based on [date-holidays](https://github.com/commenthol/date-holidays) library for creating holidays for specific set of countries.
  - Everything is configured via objects in `config.js`. The only thing that can be set outside is an environment variable to set whether the script should write logs to the console.
  - Library supports 197 countries and regions, but I only chose 40 that are currently in the app. If needed, extending it is just modifying a variable with countries in `config.js`.
    - If we'd like to add region support, e.g. different states in the USA or Australia, we'd need to modify the script because I haven't considered this case.
  - GitHub action that creates pull request with updated holiday files. It should run automatically every 1st December (I hope it will work fine, I've never set cron running so rarely) or can be triggered manually. 
- What hasn't been done:
  - `date-holidays` unfortunately doesn't support some countries that already had holidays: India, Kazakhstan, Pakistan and Sri Lanka. I've added TODO comments with links to issues on library's GH to not forget about those.
  - I haven't used `date-holidays-ical` as I wrote in the original discussion because it had dependency to an old version of `date-holidays`. Instead, I decided to use a separate library for generating ICS files.
  - Current ICS files have fixed-date holidays set as always recurring. Unfortunately, from the library I could just get entry for each specific year, so I've added it like this, so e.g. if someone currently imported holidays with Christmas, they could see Christmas on 2030-12-24. Now it will only be visible in years for which ICS are generated. 
    - If needed, it could be changed because `date-holidays` returns metadata with repeating rule. I could figure out from it which holidays have fixed dates. But the question is, whether we want to complicate that script this much.
- What I'm unsure of:
  - I didn't know for what time period I should generate holidays. Now I've set generating for current and the next year. However, I don't see any problem with creating more holidays, since these are plain text files and they shouldn't consume too much space on user devices. However, I don't know what's other users' preference.
- Additional remarks about implementation:
  -  `date-holidays` returns different kinds of holidays. I've decided to filter them, to have only public and bank holidays, since I've noticed that some countries don't have other filled out. Also, for me, it would be confusing to have also other holidays, but I can easily change it (again, it's just a variable in `config.js`)
  - `uid` of each event is reproducible (sha-256 of country and date) to avoid duplicate entries after adding holidays in recurring years. I only hope that there isn't any country that has multiple holidays on the same date, so they don't get filtered out. I decided not to use the name of the holiday because some are localized, some not, so I don't consider the name as something constant.
  - I've tested GitHub action, but on a separate, private repository. I don't know why I couldn't run any workflow on a fork, so I think you just have to believe me it works.
  - Remark for JS devs: I know that we should use `npm ci` to install packages on CI and commit `package-lock.json`, but I specially omitted it, so the script always fetches the latest version of the library. I can only hope it won't end like [faker](https://www.thegingerviking.com/the-right-to-delete-fakerjs-fragile-nature-open-source/) and there won't be any major API changes between minor versions (there shouldn't, but who knows...).
  - Probably it would be nice to add somewhere information (readme, contribution guidelines, faq, wherever) that holidays are autogenerated, so they can have errors or missing localization. Maybe also with the link to the library's repository: https://github.com/commenthol/date-holidays, so they would raise problems or change localization there.
  - I've added some comments to the code so it can be easily understood by others, but if something will be wrong after merging, feel free to ping me.

#### Fixes the following issue(s)
<!-- Prefix issues with "Fixes" so that GitHub closes them when the PR is merged (note that each "Fixes #" should be in its own item). Also add any other relevant links. -->
- Fixes #11 ⇽ I know it was a discussion, not an issue, but I'm curious if GitHub will do anything with this link.

#### Acknowledgement
- [X] I read the [contribution guidelines](https://github.com/FossifyOrg/Calendar/blob/master/CONTRIBUTING.md).

pixincreate commented on .github/workflows/holiday-generator/config.js:
```diff
@@ -0,0 +1,84 @@
+
+export const ICS_PATH = join(cwd(), "../../../app/src/main/assets");
+
+// country codes from https://github.com/commenthol/date-holidays?tab=readme-ov-file#supported-countries-states-regions
```
https://www.thunderbird.net/en-US/calendar/holidays
https://www.officeholidays.com/subscribe/

you could also parse from any one from above? Asking so because I found them to up-to-date (had been using officeholidays well over 2 years)

tswistak replied:
officeholidays is copyrighted, so we can't use it without violating the license. Thunderbird's calendar is AFAIR licensed as CC BY-SA, so we'd need to write that they are authors of these calendars and distribute them on the same license. TBH, I don't know if and how CC BY-SA is compatible with GPLv3, so using MIT-licensed date-holidays was the easiest way license-wise.

pixincreate replied:
Got it, did not think about licensing though 😅

naveensingh commented on .github/workflows/holiday-generator.yml:
```diff
@@ -0,0 +1,31 @@
+        name: Generate ICS
+        runs-on: ubuntu-latest
+        steps:
+            - uses: actions/checkout@master
```
Shouldn't we use `actions/checkout@v4`? 

The [`master`](https://github.com/actions/checkout/tree/master) branch doesn't seem to be updated anymore as the last commit was 4 years ago. Or am I missing something here?

tswistak replied:
Well, the only reason why I've used `master` is because I've copied it from another of my GitHub Actions 😄. I think I can safely change it to `v4`, there shouldn't be any breaking changes, especially that its only purpose is to check out the repo.

naveensingh commented on .github/workflows/holiday-generator.yml:
```diff
@@ -0,0 +1,31 @@
+                node-version: 20
+            - name: Run script
+              env:
+                SHOULD_LOG: true
```
This should be `LOG_ENABLED` as per line config.js:4

tswistak commented:
There's a one thing that came into my mind - some countries have multiple languages, e.g. https://github.com/commenthol/date-holidays/blob/master/data/countries/BE.yaml. Now, script works in a way, that it takes the first language (e.g. here it will take French for Belgium). Would we like to support multiple languages, and if yes, how?

naveensingh commented:
> If needed, it could be changed because date-holidays returns metadata with repeating rule. I could figure out from it which holidays have fixed dates. But the question is, whether we want to complicate that script this much.

Google Calendar on Android doesn't seem to show fixed-date holidays for the next year but we already have some requests for this:

- https://github.com/FossifyOrg/Calendar/issues/84
- https://github.com/FossifyOrg/Calendar/issues/149

We should at least 'try' to parse and add the holidays with fixed dates.

> date-holidays returns different kinds of holidays. I've decided to filter them, to have only public and bank holidays, since I've noticed that some countries don't have other filled out.

Ideally, we should import the other holidays and then allow filtering them out in the app itself but let's go with `public` and `bank` and we'll see how it works. Related feature request: https://github.com/FossifyOrg/Calendar/issues/95

> There's a one thing that came into my mind - some countries have multiple languages, e.g. https://github.com/commenthol/date-holidays/blob/master/data/countries/BE.yaml.

That's been requested as well: https://github.com/FossifyOrg/Calendar/discussions/167

> Now, script works in a way, that it takes the first language (e.g. here it will take French for Belgium). Would we like to support multiple languages, and if yes, how?

I assume the first language is the local default for countries with holidays in multiple languages? If yes, let's go with that. I'll consider adding support for multiple languages/regions some day.

tswistak commented:
Thanks for the review! Later this week, I will write a parser for fixed-date holidays.

naveensingh commented:
Ok, thank you.

tswistak commented:
@naveensingh I've found some time today and added a parser for fixed date holidays. Here's a sample ICS file that the script now creates: [unitedkingdom.ics.zip](https://github.com/FossifyOrg/Calendar/files/14593096/unitedkingdom.ics.zip).

Some tech notes:
- I've modified a bit how IDs are produced - now I also consider a repeating rule, not only the date, to ensure that fixed-date holidays won't duplicate in the app. It may also address the potential problem I described originally: `I only hope that there isn't any country that has multiple holidays on the same date, so they don't get filtered out.`
- Fixed-date holidays are generated starting from 1970, but it can be easily modified in config file.

naveensingh commented:
@tswistak Thanks for the quick work and detailed changelogs! I'll check it out tomorrow (next 10-15hrs)...

naveensingh commented on .github/workflows/holiday-generator/main.js:
```diff
@@ -0,0 +1,125 @@
+function isFixedDate(rule) {
+    return /^\d\d-\d\d$/.test(rule);
+}
+
```
This won't work for some fixed dates. The data that `getHolidays` returns doesn't strictly follow the `MM-DD` rule. For example, new year's rule:

![image](https://github.com/FossifyOrg/Calendar/assets/36371707/5e1ab0c2-8e3f-4ee5-ab01-d04b3f7fcf4f)

I think we can safely discard these conditions as long as the `MM-DD` pattern exists. There are more human-readable rules like `4th thursday in November` that we obviously can not parse without complicating things too much.

tswistak replied:
Seeing the [specification](https://github.com/commenthol/date-holidays/blob/master/docs/specification.md#fixed-date), I'd change the regex to check if the pattern is `MM-DD` or `MM-DD and ...`. Too many different exceptions to handle, these two are the most straightforward.

naveensingh replied:
I 'think' we can get rid of the `and` too and match anything like `MM-DD.*`.

tswistak replied:
In the USA's holidays, we have a rule `04-15 if friday then next monday if saturday,sunday then next tuesday`. If we'd simplify regex this way, sometimes date will be wrong.

naveensingh replied:
I couldn't get Google Calendar to show holidays in the future years but I see the start and end dates are different in date-holidays when the provided conditions are applicable. Let's keep `and` then, we'll worry about it after some feedback from users.

naveensingh commented on .github/workflows/holiday-generator/main.js:
```diff
@@ -0,0 +1,125 @@
+                    uid,
+                    start: getDateArray(x.start),
+                    end: getDateArray(x.end),
+                    recurrenceRule: "FREQ=YEARLY",
```
We should probably append us `FREQ=YEARLY;BYMONTH=$month;BYMONTHDAY=$dayOfMonth`

tswistak replied:
I've checked current ICS files and there are either `RRULE:FREQ=YEARLY` or `RRULE:FREQ=YEARLY;INTERVAL=1`.

tswistak replied:
I can modify this way, but I don't know if it's necessary. The current way seems to work fine, but if there's any advantage in specifying exact month and day, I can change it. 

I'm rather thinking about specifying `INTERVAL=1` every time because after checking random ICS files in Calendar, only `poland.ics` has no interval defined (but it's adding dates correctly).

naveensingh replied:
There's only one advantage that it's more explicit but I tested importing a couple of events with `FREQ=YEARLY` and it seems to work fine so we can leave it as it is. It's not like the ICS files will be used outside the app.

The `FREQ=YEARLY;INTERVAL=X` means the event should repeat every 1 year. Not needed as far as I know.

naveensingh commented:
Everything else looks good to me! I'm ready to merge this if you are.

tswistak commented:
I'm also ready. I'll try to take a look sometimes if there are issues about holidays and, if needed, will fix or modify the generator.

naveensingh commented:
Ok, thanks :)

tswistak commented:
I've noticed that action has failed. I forgot to mention that you have to allow GitHub actions to create pull requests. You may find it in Actions ⇾ General ⇾ Workflow permissions

naveensingh commented:
Yep, allowed it. Now running it again...

maverick74 commented:
Hi @naveensingh @tswistak 

Sorry for picking up an old issue!

There's an Holliday missing in Portugal: January 1st

also, h think it would be nice if we could generate Hollidays for, say, the next 5 years or even more...

thanks
